src/dom_types.rs

```diff
@@ -653,17 +662,80 @@ impl<Ms: 'static, OtherMs: 'static> MessageMapper<Ms, OtherMs> for Vec<El<Ms>> {
 }
 
 impl<Ms> ElContainer<Ms> for El<Ms> {
-    fn els(self) -> Vec<El<Ms>> {
-        vec![self]
+    fn els(self) -> Vec<Node<Ms>> {
+        vec![Node::Element(self)]
     }
 }
 
 impl<Ms> ElContainer<Ms> for Vec<El<Ms>> {
-    fn els(self) -> Vec<El<Ms>> {
+    fn els(self) -> Vec<Node<Ms>> {
+        self.into_iter().map(Node::Element).collect()
+    }
+}
+
+impl<Ms: 'static> ElContainer<Ms> for Node<Ms> {
+    fn els(self) -> Vec<Node<Ms>> {
+        vec![self]
+    }
+}
+
+impl<Ms: 'static> ElContainer<Ms> for Vec<Node<Ms>> {
+    fn els(self) -> Vec<Node<Ms>> {
         self
     }
 }
 
+/// For representing text nodes.
+#[derive(Clone, Debug)]
+pub struct Text {
+    pub text: String,
+    pub node_ws: Option<web_sys::Node>,
+}
+
+impl PartialEq for Text {
+    fn eq(&self, other: &Self) -> bool {
+        self.text == other.text
+    }
+}
+
+impl Text {
+    pub fn new(text: String) -> Self {
+        Self {
+            text,
+            node_ws: None,
+        }
+    }
+}
+
+/// An component in our virtual DOM. Related to, but different from
+/// [DOM Nodes](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType)
+#[derive(Clone, Debug, PartialEq)]
+pub enum Node<Ms: 'static> {
+    Element(El<Ms>),
+    //    Svg(El<Ms>),  // May be best to handle using namespace field on El
+    Text(Text),
+    Empty,
+}
```

Comment:
Idea:
Introduce something like `VirtualNode` - its variants wouldn't have `node_ws` and `ref_`(?). And remove `Option<..>` from `Node`'s `node_ws` and `ref_`. So Rust's type system would help to reduce errors and boiler plate in patch algorithm (`take`s and `expects`). It would be also faster because it wouldn't be necessary to check `node_ws` present in runtime.
But I don't know how much work it would mean and I'm not sure if it wouldn't make code less readable / create some ergonomics problems.

Reply:
I can see how that could simplify patching logic - we currently have to make sure certain steps are completed (assigning `web_sys::Node`s, attaching `Listener`s etc), and if they're not at the right times, we get panics or incorrect renders. Need to think on how this would be implemented.

Reply:
Do you think we can implement it in another PR?

Reply:
Yep

Reply:
https://github.com/David-OConnor/seed/issues/180
**RESOLVED**